ENT-94: Enroll students to a program

e-kolpakov opened the pull request:
**Description:** This PR adds ability to enroll students into a program

**JIRA:** [ENT-94](https://openedx.atlassian.net/browse/ENT-94)

**Dependencies:** no dependencies so far

**Testing instructions:**

First, prepare the environment:
1. Create Enterprise Customer (or reuse existing)
2. Create a couple of courses (or reuse existing)
3. Launch LMS and course-discovery devstack, make sure they are working together nicely.
4. In course catalog create course runs for courses selected for this test session. Add seats to those courses (+ make sure there are courses with multiple seat types and there's at least one combination of courses that have at least one common seat type).
5. Create program(s) and include some combinations of courses into each.

Testing instructions:
1. Attempt adding single user and enrolling it into a program by UUID:
1.a. Go to course discovery admin and note program UUID (or send a GET to `course-discory/api/v1/programs` and pick one from the response)
1.b. Go to "Manage Learners" view
1.c. Fill in user email
1.d. Paste program UUID into a field
1.e. Choose one of the common enrollment modes
1.f. Submit form
1.g. Observe: user is added to corresponding list of users and (will be) enrolled into all courses in the program with the chosen enrollment mode
2. Attempt adding single user and enrolling it into a program by title - same as (1), but in step d use program title instead of ID.
3. Same as 1, but use bulk students upload - all students must be enrolled in all courses in chosen program.
4. Same as 2, but use bulk students upload.
5. Check validation:
5.a. If program was not found neither by UUID nor by name - form should display sensible error and not enroll/link student(s)
5.b. If chosen mode is not available across all selected course runs - form should display sensible error and not enroll/link student(s) and also list what common modes are available for courses in the program.
5.c. If LMS API or course discovery API is not available - it is going to crash when submitting the form (behavior not affected by this PR).
6. Make sure JS interactions work fine:
6.1. When non empty string is typed/pasted into course ID, program ID input becomes disabled and after 300 milliseconds after course modes are loaded (if user continues to type it waits) and put into course mode select.
6.2. When non empty string is typed/pasted into program ID, course ID input becomes disabled and after 300 milliseconds after default course modes are put into course mode select.
6.3. When course ID or program ID is emptied, other input becomes available immediately.

**Reviewers:**
- [x] @haikuginger 
- [x] @asadiqbal08 
- [x] @saleem-latif

CC @mattdrayer 

**Merge checklist:**
- [x] All reviewers approved
- [ ] CI build is green
- [x] Version bumped
- [x] Changelog record added
- [x] Commits are (reasonably) squashed
- [x] Translations are updated
- [x] PR author is listed in AUTHORS

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is finished.
- [ ] Delete working branch (if not needed anymore)

**Author concerns:**

`EnterpriseCustomerManageLearnersView` has become too big and performs too much actions: manages form, links users to EnterpriseCustomer, enrolls them into courses, etc. It makes it harder and harder to test. Might be good to start pulling it apart, or at least to split the logic of parsing user input and actually performing actions on that input (enrolling/linking students) into a separate objects.

Also, `ManageLearnersForm` starts to become too big too - might be a good place to stop and re-think the enrollment+linking approach.

edx-webhook commented:
Thanks for the pull request, @e-kolpakov! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-enterprise&number=29"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

haikuginger commented on enterprise/admin/utils.py:
```diff
@@ -112,3 +130,18 @@ def validate_email_to_link(email, raw_email=None, message_template=None, ignore_
+def get_course_runs_from_program(program):
+    """
+    Return course runs from program data.
+
```
EdX has been asking for arguments listed in the docstrings of new and updated methods - could you add that here and wherever else is relevant?

haikuginger commented on enterprise/admin/utils.py:
```diff
@@ -112,3 +130,18 @@ def validate_email_to_link(email, raw_email=None, message_template=None, ignore_
+    Returns:
+        set: course runs in given program
+    """
+    return set(
```
The logic here is a little hard to follow; I think it might be better to break this out into a set of nested clauses rather than a comprehension.

e-kolpakov replied:
@haikuginger That's just a two-level for comprehension with filter, nothin' serious :)

haikuginger commented on enterprise/course_catalog_api.py:
```diff
@@ -22,50 +24,93 @@
-    pass
+    def __init__(self, user):
+        """
+        Create an Enrollment API client, authenticated with the API token from Django settings.
```
s/Enrollment/Course Catalog/

haikuginger commented on enterprise/course_catalog_api.py:
```diff
@@ -22,50 +24,93 @@
+        """
+        all_programs = self._load_data('programs')
+        try:
+            return next(program for program in all_programs if program.get('title') == program_title)
```
Should we return a single program in a scenario where multiple programs match the same title? Or should we raise an exception to complain about the ambiguity?

Better yet, we might want to yield each matching program from here, and then raise later on if either none of the programs are eligible for enrollment, or if multiple programs are eligible for enrollment.

e-kolpakov replied:
@haikuginger I've added some protection against it (haven't pushed it yet), but title is a primary/unique key for course catalog program:

```
IntegrityError at /admin/course_metadata/program/add/
(1062, "Duplicate entry 'Program1' for key 'title'")
```

haikuginger commented on test_utils/fake_catalog_api.py:
```diff
@@ -0,0 +1,263 @@
+    Fake implementation returning program by title.
+    """
+    try:
+        return next(response for response in FAKE_PROGRAM_RESPONSES.values() if response["title"] == program_title)
```
Note that we'll need to change the implementation here to match the implementation in the actual code if that modification is made.

e-kolpakov replied:
@haikuginger it is a good point, but it might make more sense to keep it as is, since it is a fake method - it is supposed to be simpler than actual implementation. Also, it won't have such problem (at least for now), since the only two programs avaiable to it have different titles.

haikuginger commented on tests/admin/test_forms.py:
```diff
@@ -19,29 +20,51 @@
 FAKER = FakerFactory.create()
 
 
+class TestWithCourseCatalogApiMixin(object):
```
To make this a proper mixin, let's do `self.catalog_api = self._make_catalog_api()` in this class's `setUp` method. Or, even inline that code into this class's `setUp`.

haikuginger commented on tests/admin/test_forms.py:
```diff
@@ -19,29 +20,51 @@
 @mark.django_db
 @ddt.ddt
-class TestManageLearnersForm(unittest.TestCase):
+class TestManageLearnersForm(unittest.TestCase, TestWithCourseCatalogApiMixin):
```
Let's place `TestWithCourseCatalogApiMixin` before `unittest.TestCase` here. That way, its `setUp` method will be called.

haikuginger commented on tests/admin/test_forms.py:
```diff
@@ -222,9 +245,114 @@ def test_clean_valid_course_invalid_mode(self, enrollment_client):
 
 @mark.django_db
-class TestEnterpriseCustomerAdminForm(unittest.TestCase):
+class TestEnterpriseCustomerAdminForm(unittest.TestCase, TestWithCourseCatalogApiMixin):
```
Mixin changes as above.

haikuginger commented on tests/test_course_catalog_api.py:
```diff
@@ -0,0 +1,200 @@
+    @ddt.data(
+        "course-v1:JediAcademy+AppliedTelekinesis+T1",
+        "course-v1:TrantorAcademy+Psychohistory101+T1",
+        "course-v1:StarfleetAcademy+HyperspaceNavigation+T2337",
```
s/Hyperspace/Warpspeed/

😜

e-kolpakov replied:
@haikuginger Aw crap, right :)

haikuginger commented:
@e-kolpakov, behavior is looking pretty good in general! One thing that would be great - you'll note that on the current master, we have javascript that disables the Mode dropdown when a valid course mode hasn't been entered. As a result, it's currently impossible to select a mode via the UI. It would be great if we could dynamically load the valid program modes into the list when a program ID or name is entered, but failing that, we should at least set the dropdown to have the default values and be enabled.

e-kolpakov commented:
@haikuginger I've addressed most of your comments - please take another look.

I plan to add tests for the new frontend functionality tomorrow, but please take a moment to look at it and comment on usability.

mattdrayer commented on CHANGELOG.rst:
```diff
@@ -14,6 +14,15 @@ Change Log
 Unreleased
 ~~~~~~~~~~
 
+[0.14.0] - 2017-01-15
```
We've got another `0.14.0` in the works [here](https://github.com/edx/edx-enterprise/pull/30/files#diff-de0a9b67ffe2874a076b5fdb15bad484R17), just FYI to you and @saleem-latif

e-kolpakov replied:
@mattdrayer thanks. No one knows for sure which PR will be merged first, so it makes little sense to try and predict what version should one's PR have

mattdrayer commented on enterprise/admin/forms.py:
```diff
@@ -14,10 +14,12 @@
 from enterprise import utils
-from enterprise.admin.utils import ValidationMessages, email_or_username__to__email, validate_email_to_link
-from enterprise.course_catalog_api import get_all_catalogs
+from enterprise.admin.utils import (ProgramStatuses, ValidationMessages, email_or_username__to__email,
```
At what point do we just import the entire `enterprise.admin.utils` package?

e-kolpakov replied:
@mattdrayer if you mean `from admin.utils import *` - probably never as pylint will complain. If you mean just `import admin.utils` and prefixing utils members - probably someday. But the better approach might be to split utils into more granular modules or (to the opposite) impose some classes to contain logically related utils.

mattdrayer commented on enterprise/admin/forms.py:
```diff
@@ -35,6 +37,10 @@ class ManageLearnersForm(forms.Form):
         help_text=_("Provide a course ID if enrollment is desired."),
     )
+    program = forms.CharField(
+        label=_("Or enroll these learners in this program"), required=False,
```
Is this either/or selection (I guess it's actually Also/Or) enforced in the code?

e-kolpakov replied:
@mattdrayer only on frontend so far, but it's just because I didn't have time to implement it yesterday.

mattdrayer commented on enterprise/admin/forms.py:
```diff
@@ -65,13 +71,25 @@ class Fields(object):
 
+    def __init__(self, *args, **kwargs):
+        """
+        Initializes
```
Would be great to have a more descriptive docstring here

e-kolpakov replied:
@mattdrayer sure. Looks like I've started typing a docstring, got distracted and forgot about it :)

mattdrayer commented on enterprise/admin/forms.py:
```diff
@@ -65,13 +71,25 @@ class Fields(object):
+        Initializes
+
+        Arguments:
+            user(django.contrib.auth.models.User): current user
```
Nit: Insert a space between user and parenthesis -> `user (django...`

mattdrayer commented on enterprise/admin/forms.py:
```diff
@@ -103,6 +121,37 @@ def clean_course(self):
+        if not program_id:
+            return None
+
+        try:
```
I'd prefer to see specific try/excepts versus wrapping all of the work we want to protect in one big try with multiple excepts at the end.

mattdrayer replied:
@e-kolpakov -- just checking -- are you planning on addressing this suggestion or leaving as-is?

mattdrayer commented on enterprise/admin/forms.py:
```diff
@@ -144,7 +201,28 @@ def clean(self):
+                raise ValidationError(ValidationMessages.COURSE_MODE_NOT_AVAILABLE.format(
+                    mode=course_mode, program_title=program.get("title"), modes=", ".join(available_modes)
+                ))
+        except (HttpClientError, HttpServerError):
```
This except can be moved closer to the location being protected.

mattdrayer commented on enterprise/admin/forms.py:
```diff
@@ -185,7 +263,8 @@ def get_catalog_options(self):
         for the `choices` argument to a ChoiceField.
         """
-        catalogs = get_all_catalogs(self.user)
+        catalog_api = CourseCatalogApiClient(self.user)
```
Do we need to protect against network/communication errors here?

mattdrayer replied:
Checking on this one as well -- `try/except`?

e-kolpakov replied:
@mattdrayer Question to @saleem-latif as author of this code - I only switched it to use `CourseCatalogApiClient` as part of refactoring.

mattdrayer commented on enterprise/admin/utils.py:
```diff
@@ -28,8 +38,19 @@ class ValidationMessages(object):
+    COURSE_MODE_NOT_AVAILABLE = _(
+        "Course mode {mode} is not available for all courses in program {program_title}. Available modes are {modes}."
+    )
+    FAILED_TO_OBTAIN_MODES = _("Failed to obtain available course modes for program {program_title}")
```
`FAILED_TO_OBTAIN_PROGRAM_COURSE_MODES` ?

mattdrayer commented on enterprise/admin/utils.py:
```diff
@@ -28,8 +38,19 @@ class ValidationMessages(object):
-    COURSE_WITHOUT_COURSE_MODE = _("Please select a course enrollment mode for the given course.")
+    COURSE_WITHOUT_COURSE_MODE = _("Please select a course enrollment mode for the given course/program.")
     INVALID_COURSE_ID = _("Could not retrieve details for the course ID {course_id}. Please specify a valid ID.")
+    INVALID_PROGRAM_ID = _(
```
I go back and forth on whether it's better to alphabetize or group by concept -- we need to do it one way or the other, however, and alphabetical is the current convention, so maybe we should stick with that.

mattdrayer commented on enterprise/admin/views.py:
```diff
@@ -322,13 +323,19 @@ def post(self, request, customer_uuid):
             if course_details:
+                course_ids.append(course_details['course_id'])
+
+            program_detais = manage_learners_form.cleaned_data.get(ManageLearnersForm.Fields.PROGRAM)
```
program_detais -> program_details

mattdrayer replied:
Should we still do this work if there are course_details available from above?

mattdrayer replied:
ah, I think I see now, nevermind

mattdrayer commented on enterprise/course_catalog_api.py:
```diff
@@ -22,50 +24,126 @@
+                available_course_modes &= course_run_modes
+
+            if not available_course_modes:
+                return available_course_modes
```
If there are three course_runs and the first two do not have course_run_modes, but the third one does, will that information be missed due to this return statement?

e-kolpakov replied:
@mattdrayer in this scenario, since the first two does not have run modes the right result would be an empty set. I'll add a test for this case.

mattdrayer replied:
Okay -- it might be helpful for future developers if we explain what is going on here with some code comments.

mattdrayer commented on enterprise/course_catalog_api.py:
```diff
@@ -22,50 +24,126 @@
+            if not available_course_modes:
+                return available_course_modes
+
+        return available_course_modes
```
I'm not really seeing any sort of commonality/reduction process here -- it seems like we actually mean "all defined modes", versus what is actually common across all course runs?

e-kolpakov replied:
@mattdrayer `available_course_modes` is a set. `&=` on line 126 is apply `&` to lhs and rhs and store in lhs. `&` on sets is defined as intersection of the operands, so it does the filtering.

There should be tests to check that this returns only modes available across all course runs.

mattdrayer replied:
Cool, thanks for the explanation, I had a hard time finding a solid description of `&=` on my own.  I think I get it now.  So ultimately the end result will contain only those modes which appear across all course runs.

mattdrayer commented on enterprise/utils.py:
```diff
@@ -42,6 +42,33 @@ def __init__(self, *args, **kwargs):
+    """
+
+
+class MultipleProgramMatch(CourseCatalogApiError):
```
Seems like this class name should end in either Error or Exception

e-kolpakov commented:
@haikuginger @mattdrayer Please take another look

haikuginger commented on enterprise/static/enterprise/admin/manage_learners.js:
```diff
@@ -70,24 +83,43 @@ function loadPage() {
+    };
+    
+    var defaultModes = [
+        {slug: "", name: "---------"},
```
Nit: Should these be marked for i18n?

haikuginger commented:
:+1:

- [x] I tested this: I was able to enroll users in programs
- [x] I read through the code
- [x] I checked for accessibility issues
- [x] ~~Includes documentation~~ N/A

e-kolpakov commented:
@asadiqbal08 @saleem-latif @mattdrayer Ready for you reviev.

mattdrayer commented on enterprise/admin/views.py:
```diff
@@ -322,13 +323,19 @@ def post(self, request, customer_uuid):
+
             if course_details:
+                course_ids.append(course_details['course_id'])
+            elif program_details:
```
Is it ever possible for both course_details and program_details to exist simultaneously?

e-kolpakov replied:
@mattdrayer No, now it is prevented by frontend and form validation

mattdrayer commented:
Just a couple things remaining for me -- @asadiqbal08 and/or @saleem-latif please take a look as well.

e-kolpakov commented:
@mattdrayer @saleem-latif @asadiqbal08 rebasing from f01d687704ad041b622655955fffe20809dab896 onto master

e-kolpakov commented on enterprise/__init__.py:
```diff
@@ -4,6 +4,6 @@
 from __future__ import absolute_import, unicode_literals
 
-__version__ = "0.13.0"
+__version__ = "0.14.0"
```
@asadiqbal08 I'm still rebasing :)

e-kolpakov commented:
@asadiqbal08 Rebased

asadiqbal08 commented:
@e-kolpakov sorry I could not see the version update in `edx-enterprise/enterprise/__init__.py`.

e-kolpakov commented:
@asadiqbal08 Good catch - I've missed that one.

e-kolpakov commented:
@asadiqbal08 bumped

asadiqbal08 commented:
👍

e-kolpakov commented:
@mattdrayer Final check before merging - good to go>

mattdrayer commented:
:+1: :ship:

e-kolpakov commented:
@mattdrayer :rocket:  --> :artificial_satellite:  --> :ship:  --> :boom:
